Allow reflection for native serialization

vsevel opened the pull request:
Fixes https://github.com/quarkusio/quarkus/issues/19711

I have assumed that for reflection we needed `constructor=true` and we honor annotation parameters `methods` and `fields` (just like we do for the non serialization case).
I kept `finalFieldsWritable` and `weak` false. `weak` and `serialization` appear to be exclusive.
I am not sure `finalFieldsWritable` makes sense with `serialization`.

/cc @jaikiran @JiriOndrusek @stuartwdouglas

vsevel committed:
Allow reflection for native serialization

jaikiran commented:
Hello @vsevel, this looks OK to me at first glance. However, I haven't looked at the deserialization code in JDK and what other reflection access other than constructors might be needed. I think what would be good and definitive is a simple test case (in native mode) which reproduces the original issue and then verifies this change to make sure the serialization (and internally reflection) works as expected. We already have native tests which run REST endpoints, so I guess those existing ones can be enhanced to test this usecase, maybe?

vsevel committed:
test

vsevel commented:
hello @jaikiran I have added a list field in the `SomeSerializationObject` class used in the `CoreSerializationInGraalITCase`. unfortunately the test fails. I searched quite a bit but could not figure out what is going on. I have added some logging in the `ReflectiveClassBuildItem`, `NativeImageAutoFeatureStep` and `RegisterForReflectionBuildStep` to make sure the annotation was properly processed, and produced the appropriate build items.

```
@RegisterForReflection(targets = { List.class, ArrayList.class, String.class }, serialization = true)
public class SerializationConfig {
```

I decompiled also the fat jar, and checked that the `AutoFeature` had the following:

```
    private static void registerClass1(BeforeAnalysisAccess var0) {
        try {
            ClassLoader var1 = Thread.currentThread().getContextClassLoader();
            Class var2 = Class.forName("java.util.ArrayList", (boolean)0, var1);
            Constructor[] var4 = var2.getDeclaredConstructors();
            Method[] var5 = var2.getDeclaredMethods();
            Field[] var8 = var2.getDeclaredFields();
            Class[] var3 = new Class[]{var2};
            RuntimeReflection.register(var3);
            RuntimeReflection.register((Executable[])var4);
            RuntimeReflection.register((Executable[])var5);
            Version var6 = Version.getCurrent();
            int[] var7 = new int[]{21};
            if (var6.compareTo(var7) < 0) {
                RuntimeReflection.register((boolean)0, var8);
            } else {
                RuntimeReflection.register((boolean)0, (boolean)1, var8);
            }

            registerSerializationForClass(var2);
        } catch (Throwable var9) {
        }

    }
```

Note that constructors get registered for the `ArrayList`.

Yet, the test keeps failing with:

```
[ERROR]   CoreSerializationInGraalITCase.testEntitySerializationFromServlet:16 1 expectation failed.
Response body doesn't match expectation.
Expected: is "OK"
  Actual: java.io.InvalidClassException: java.util.ArrayList; no valid constructor
        java.io.InvalidClassException: java.util.ArrayList; no valid constructor
        at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:159)
        at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:875)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2170)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:493)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:451)
        at io.quarkus.it.corestuff.SerializationTestEndpoint.reflectiveSetterInvoke(SerializationTestEndpoint.java:53)
```

Could this be an issue in graalvm itself?

zakkak commented:
> Could this be an issue in graalvm itself?

No, that's not an issue in GraalVM. The issue appears to be Quarkus passing the wrong argument to `com.oracle.svm.reflect.serialize.hosted.SerializationFeature.addReflections` in https://github.com/quarkusio/quarkus/blob/550e59a4faa97fda1a66c466b9b0c77582706d21/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java#L672-L674

I created #20167 to fix it.

zakkak commented:
Closing, since https://github.com/quarkusio/quarkus/pull/20167, which fixes the immediate issue, is now merged.
I have also created https://github.com/quarkusio/quarkus/issues/20300 to request a new annotation for registering classes for serialization.
